### PR TITLE
set scrape wanted

### PR DESCRIPTION
**Pull request by yingjieMiao**

### CODE REVIEW

**Comment by yingjieMiao** [comment]
@leogrim 

L97 - 103 in normUriRepo was doing the samething as the old code.
(In the old code, if a uri is new, it's first created, with active state, and then, we set srapeInfo manually later) So (I think) we didn't miss anything to scrape.

File: kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala [hunk 1]
```diff
@@ -79,11 +79,10 @@ class UriIntegrityActor @Inject()(
...
-  private def handleScrapeInfo(newUriId: Id[NormalizedURI])(implicit session: RWSession) = {
-    scrapeInfoRepo.getByUri(newUriId) match {
-      case Some(info) if (info.state == ScrapeInfoStates.INACTIVE) => scrapeInfoRepo.save(info.withState(ScrapeInfoStates.ACTIVE))
-      case None => scrapeInfoRepo.save(ScrapeInfo(uriId = newUriId))
-      case _ =>
+  private def handleScrapeInfo(oldUri: NormalizedURI, newUri: NormalizedURI)(implicit session: RWSession) = {
+    val oldInfo = scrapeInfoRepo.getByUri(oldUri.id.get)
+    if (oldInfo.isDefined && oldInfo.get.state == ScrapeInfoStates.ACTIVE){
+      uriRepo.save(newUri.withState(NormalizedURIStates.SCRAPE_WANTED))
     }
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala:89]: We should check that newUri doesn't already have a scrape info e.g. newUri is already "Scraped" in which case we don't need to set it to "Scrape_Wanted" and touch scrape_info at all

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala:89]: So in short, if oldInfo active and newInfo is absent or inactive, save newUri with ScrapeWanted. That's it right?

Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala:89]: yeah. you are right.

File: kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala [hunk 2]
```diff
@@ -79,12 +79,12 @@ class UriIntegrityActor @Inject()(
...
   }
   
-  private def handleScrapeInfo(newUriId: Id[NormalizedURI])(implicit session: RWSession) = {
-    scrapeInfoRepo.getByUri(newUriId) match {
-      case Some(info) if (info.state == ScrapeInfoStates.INACTIVE) => scrapeInfoRepo.save(info.withState(ScrapeInfoStates.ACTIVE))
-      case None => scrapeInfoRepo.save(ScrapeInfo(uriId = newUriId))
-      case _ =>
-    }
+  private def handleScrapeInfo(oldUri: NormalizedURI, newUri: NormalizedURI)(implicit session: RWSession) = {
+    if (newUri.state == NormalizedURIStates.ACTIVE){    // otherwise, we have tried scraping. (we assume this new uri is not inactive or redirected)
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala]: @Loe

Is this better?

File: kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala [hunk 3]
```diff
@@ -124,10 +124,10 @@ class UriIntegrityActor @Inject()(
       log.info(s"migrating url ${url.id} to new uri: ${newUriId}")
       
       urlRepo.save(url.withNormUriId(newUriId).withHistory(URLHistory(clock.now, newUriId, URLHistoryCause.MIGRATED)))
-      val newUri = uriRepo.get(newUriId)
+      val (oldUri, newUri) = (uriRepo.get(url.normalizedUriId), uriRepo.get(newUriId))
       if (newUri.redirect.isDefined) uriRepo.save(newUri.copy(redirect = None, redirectTime = None))
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala:131]: According to this code, it's possible that the state of the uri is "Redirected", it doesn't have to be "Active" does it?

Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala:131]: Oh, I thought we clear the redirection field.

We now have a new uri state.

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala:131]: I think that what's most secure from future code changes is to check whether an active scrape info exists for the new uri, this is the real criteria to determine what we should do, uri states may further change in the future for whatever reason.

File: kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala [hunk 4]
```diff
@@ -124,10 +124,10 @@ class UriIntegrityActor @Inject()(
       log.info(s"migrating url ${url.id} to new uri: ${newUriId}")
       
       urlRepo.save(url.withNormUriId(newUriId).withHistory(URLHistory(clock.now, newUriId, URLHistoryCause.MIGRATED)))
-      val newUri = uriRepo.get(newUriId)
-      if (newUri.redirect.isDefined) uriRepo.save(newUri.copy(redirect = None, redirectTime = None))
+      val (oldUri, newUri) = (uriRepo.get(url.normalizedUriId), uriRepo.get(newUriId))
+      if (newUri.redirect.isDefined) uriRepo.save(newUri.copy(redirect = None, redirectTime = None).withState(NormalizedURIStates.ACTIVE))
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala:130]: @leogrim 

clear redirected states

File: kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala [hunk 5]
```diff
@@ -79,10 +79,12 @@ class UriIntegrityActor @Inject()(
...
-      case Some(info) if (info.state == ScrapeInfoStates.INACTIVE) => scrapeInfoRepo.save(info.withState(ScrapeInfoStates.ACTIVE))
-      case None => scrapeInfoRepo.save(ScrapeInfo(uriId = newUriId))
+  private def handleScrapeInfo(oldUri: NormalizedURI, newUri: NormalizedURI)(implicit session: RWSession) = {
+    val (oldInfoOpt, newInfoOpt) = (scrapeInfoRepo.getByUri(oldUri.id.get), scrapeInfoRepo.getByUri(newUri.id.get))
+    (oldInfoOpt, newInfoOpt) match {
+      case (Some(oldInfo), None) if (oldInfo.state == ScrapeInfoStates.ACTIVE) => uriRepo.save(newUri.withState(NormalizedURIStates.SCRAPE_WANTED))
+      case (Some(oldInfo), Some(newInfo)) if ( oldInfo.state == ScrapeInfoStates.ACTIVE && newInfo.state == ScrapeInfoStates.INACTIVE )  =>
+        uriRepo.save(newUri.withState(NormalizedURIStates.SCRAPE_WANTED))
       case _ =>
     }
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala:89]: @leogrim 

As you suggested.

**Pull request merged**